src/grammar/trigger.rs

```diff
@@ -0,0 +1,485 @@
+// SPDX-License-Identifier: SEE LICENSE IN LICENSE.md
+// SPDX-FileCopyrightText: 2023 CYBERTEC PostgreSQL International GmbH
+// <office@cybertec.at>
+
+use crate::lexer::TokenKind;
+use crate::parser::{expect_one_of_match, Parser};
+use crate::syntax::SyntaxKind;
+
+use super::*;
+
+/// Parses a complete trigger.
+pub(crate) fn parse_trigger(p: &mut Parser) {
+    p.start(SyntaxKind::Trigger);
+    parse_header(p);
+    parse_body(p);
+    p.finish();
+}
+
+/// Parses the header of a trigger.
+fn parse_header(p: &mut Parser) {
+    p.start(SyntaxKind::TriggerHeader);
+    p.expect(TokenKind::CreateKw);
+    if p.eat(TokenKind::OrKw) {
+        p.expect(TokenKind::ReplaceKw);
+    }
+
+    p.eat_one_of(&[TokenKind::Editionable, TokenKind::NonEditionable]);
+
+    p.expect(TokenKind::TriggerKw);
+    parse_ident(p, 1..2);
+
+    expect_one_of_match!(
```

Comment:
How about we use a similar approach using a macro ([`T`](https://github.com/rust-lang/rust-analyzer/blob/master/crates/parser/src/syntax_kind/generated.rs)) to shorten the amount of text to write. 

The main idea is to have a shorthand macro:

```rust
#[macro_export]
macro_rules ! T {
    [before] => TokenKind:::BeforeKw,
    [.] => TokenKind::Dot,
    // ...
}
```

Reason is I would like to avoid implicit code that is hidden behind a macro, in the case above it's the default case which would be hidden. Using a macro would end up in something like:

```rust
match p.current() {
    T![before] | T![after] => parse_simple_dml_trigger(p),
    T![instead] => parse_instead_of_dml_trigger(p),
    T![r#for] => p.error(ParseError::Unimplemented("compound trigger".to_string())),
    _ => p.error(ParseError::UnexpectedToken(p.current())),
}
```

In the above `match` statement the variant `ParseError::UnexpectedToken(..)` is not available yet, it's just an example.

Reply:
I agree with the `T` macro, I'll create another ticket for it. Regarding the `_ => p.error` arm, the `expect_one_of_match` macro does exactly what you described, it creates the an error listing all of the possible tokens.